requirements.txt

```diff
@@ -5,6 +5,7 @@ pip>=22.3
 # Build dependencies for non-Python components
 # Do not allow NumPy 2.0.0 due to a bug with their C API that blocks the usage of the Stable ABI;
 # this bug was fixed in 2.0.1 (https://github.com/numpy/numpy/pull/26995)
+nanobind>=2.1.0
```

Comment:
Could you comment on this requirement? For example, in the pyproject.toml there is no version lower bound.

Reply:
Just another comment on the strategy we have regarding requirements / dependencies.

- `pyproject.toml`: This is the new to specify _build_ requirements for python packages, so anything that's needed in order to build our python package (i.e. the frontend) can be put in there, and it will automatically be installed in a separate environment only for the duration of the build process.
- `requirements.txt`: Since we have other components that are not Python packages (mlir, runtime), we put their _build_ requirements in this file (e.g. numpy is needed to build mlir) and tell users to install these packages before starting the build process, so this method is a little less automated than the first one. We also put other useful development tools in here like linters, formatters, testing requirements, etc. that we would developers to have.
- `setup.py`: Here we put any _runtime_ dependencies that the built Catalyst package has, enforcing their installation when a user installs Catalyst.
- documentation: A few additional requirements that are difficult to handle automatically are just referenced in the documentation, like installing a compiler.
- `doc/requirements.txt` & `benchmark/requirements.txt`: optional dependencies only used for those particular features

Ideally, we want to avoid putting the same requirement in more than one place (unless needed), and prefer "automatic" over "manual" methods.
In this case looking at nanobind for example, we'll need it for the frontend and the runtime, so putting it on both the pyproject.toml and requirements.txt is probably needed (although technically this PR might only need it in the pyproject), but I just wanted to give a rundown of the dependency handling :)